sdk/go/common/resource/plugin/langruntime_plugin.go

```diff
@@ -55,35 +56,125 @@ type langhost struct {
-		apitype.LanguagePlugin, strings.ReplaceAll(runtime, tokens.QNameDelimiter, "_"), nil, host.GetProjectPlugins())
+	attachPort, err := GetLanguageAttachPort(runtime)
 	if err != nil {
 		return nil, err
 	}
 
-	contract.Assertf(path != "", "unexpected empty path for language plugin %s", runtime)
+	var plug *plugin
+	var client pulumirpc.LanguageRuntimeClient
+	if attachPort != nil {
+		port := *attachPort
 
-	args, err := buildArgsForNewPlugin(host, info.RootDirectory(), info.Options())
-	if err != nil {
-		return nil, err
-	}
+		handshake := func(
+			ctx context.Context, bin string, prefix string, conn *grpc.ClientConn,
+		) (*pulumirpc.LanguageHandshakeResponse, error) {
+			req := &pulumirpc.LanguageHandshakeRequest{
+				EngineAddress:    host.ServerAddr(),
+				RootDirectory:    filepath.Dir(bin),
+				ProgramDirectory: filepath.Dir(bin),
+			}
+			return languageHandshake(ctx, bin, prefix, conn, req)
+		}
 
-	plug, _, err := newPlugin(ctx, workingDirectory, path, runtime,
-		apitype.LanguagePlugin, args, nil, /*env*/
-		testConnection, langRuntimePluginDialOptions(ctx, runtime))
-	if err != nil {
-		return nil, err
+		conn, _, err := dialPlugin(
+			port,
+			"pulumi-language-"+runtime,
+			runtime+" (Language Plugin)",
+			handshake,
+			langRuntimePluginDialOptions(ctx, runtime),
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		plug = &plugin{
+			Conn: conn,
+			// Nothing to kill.
+			Kill: func() error {
+				return nil
+			},
+		}
+
+		client = pulumirpc.NewLanguageRuntimeClient(plug.Conn)
+	} else {
+		path, err := workspace.GetPluginPath(
+			ctx.Diag,
+			apitype.LanguagePlugin,
+			strings.ReplaceAll(runtime, tokens.QNameDelimiter, "_"),
+			nil,
+			host.GetProjectPlugins(),
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		contract.Assertf(path != "", "unexpected empty path for language plugin %s", runtime)
+
+		args, err := buildArgsForNewPlugin(host, info.RootDirectory(), info.Options())
+		if err != nil {
+			return nil, err
+		}
+
+		plug, _, err = newPlugin(
+			ctx,
+			workingDirectory,
+			path,
+			runtime,
+			apitype.LanguagePlugin,
+			args,
+			nil, /*env*/
+			testConnection,
+			langRuntimePluginDialOptions(ctx, runtime),
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		client = pulumirpc.NewLanguageRuntimeClient(plug.Conn)
 	}
+
 	contract.Assertf(plug != nil, "unexpected nil language plugin for %s", runtime)
 
 	return &langhost{
 		ctx:     ctx,
 		runtime: runtime,
 		plug:    plug,
-		client:  pulumirpc.NewLanguageRuntimeClient(plug.Conn),
+		client:  client,
 	}, nil
 }
 
+// Checks PULUMI_DEBUG_LANGUAGES environment variable for any overrides for the
+// language identified by name. If the user has requested to attach to a live
+// language plugin, returns the port number from the env var.
+//
+// For example, `PULUMI_DEBUG_LANGUAGES=go:12345,dotnet:678` will result in 12345 for go and 678 for dotnet.
+func GetLanguageAttachPort(runtime string) (*int, error) {
```

Comment:
Can't this share the same code as for DEBUG_PROVIDERS?

Reply:
_could_ but I hesitated to (go makes me scared to do things).  Hopefully it's ok.

Reply:
done.